lms/djangoapps/commerce/utils.py

```diff
@@ -109,3 +117,247 @@ def upgrade_url(self, user, course_key):
+        log.exception(
+            "Unexpected exception while attempting to initiate refund for user [%s], "
+            "course entitlement [%s] message: [%s]",
+            course_entitlement.user.id,
+            course_entitlement.uuid,
+            str(exc)
+        )
+        return False
+
+    if refund_ids:
+        log.info(
+            'Refund successfully opened for user [%s], course entitlement [%s]: %r',
+            enrollee.username,
+            entitlement_uuid,
+            refund_ids,
+        )
+
+        return _process_refund(
+            refund_ids=refund_ids,
+            api_client=api_client,
+            course_product=course_entitlement
+        )
+    else:
+        log.info('No refund opened for user [%s], course entitlement [%s]', enrollee.id, entitlement_uuid)
+        return True
+
+
+def refund_seat(course_enrollment):
+    """
+    Attempt to initiate a refund for any orders associated with the seat being unenrolled,
+    using the commerce service.
+
+    Arguments:
+        course_enrollment (CourseEnrollment): a student enrollment
+
+    Returns:
+        A list of the external service's IDs for any refunds that were initiated
+            (may be empty).
+
+    Raises:
+        exceptions.SlumberBaseException: for any unhandled HTTP error during communication with the E-Commerce Service.
+        exceptions.Timeout: if the attempt to reach the commerce service timed out.
+    """
+    User = get_user_model()  # pylint:disable=invalid-name
+    course_key_str = unicode(course_enrollment.course_id)
+    enrollee = course_enrollment.user
+
+    service_user = User.objects.get(username=settings.ECOMMERCE_SERVICE_WORKER_USERNAME)
+    api_client = ecommerce_api_client(service_user)
+
+    log.info('Attempting to create a refund for user [%s], course [%s]...', enrollee.id, course_key_str)
+
+    refund_ids = api_client.refunds.post({'course_id': course_key_str, 'username': enrollee.username})
+
+    if refund_ids:
+        log.info('Refund successfully opened for user [%s], course [%s]: %r', enrollee.id, course_key_str, refund_ids)
+
+        _process_refund(
+            refund_ids=refund_ids,
+            api_client=api_client,
+            course_product=course_enrollment,
+        )
+    else:
+        log.info('No refund opened for user [%s], course [%s]', enrollee.id, course_key_str)
+
+    return refund_ids
+
+
+def _process_refund(refund_ids, api_client, course_product):
+    """
+    Helper method to process a refund for a given course_product
+
+    Returns:
+        bool: True if the refund process was successful, False if there are any Errors that are not handled
+    """
+    config = CommerceConfiguration.current()
+
+    if config.enable_automatic_refund_approval:
+        refunds_requiring_approval = []
+
+        for refund_id in refund_ids:
+            try:
+                # NOTE: Approve payment only because the user has already been unenrolled. Additionally, this
```

Comment:
NOTE: At this point, for entitlements, the entitlement would NOT have been unenrolled/revoked yet.

Reply:
I will update this comment.

Reply:
With the changes in order if operation this comment is true again. I will still try to clean it up some.